### ISSUE TITLE
Fix .clang_format to not generate mixed indent.

### ISSUE THREAD

**Issue by Parakleta** [civetweb/civetweb#207]
The current `.clang_format` rules are generating mixed indent, which is fairly universally considered bad coding style.  It'll generate a hideous diff when it's implemented but I think it's something worth fixing.

The simplest change is just to remove the `UseTab:` line and fall back to the LLVM default style (or explicitly set to `None`).

I would also recommend adding `AlwaysBreakAfterDefinitionReturnType: TopLevel` to bring consistency to the function definitions (this style is currently mixed).

I am in the process of trying to refactor the code to break out the SSL into a clean interface to allow the substitution of other libraries, specifically to facilitate the optional migration to the LibreSLL/LibTLS library, and I always audit code style before refactoring.  It's easier to start with a clean consistent base.

Do you have a preference for implicit vs explicit null and zero testing.  This is currently mixed in the code and I intend to generate a patch to bring it to a consistent style, but I have no preference and the code seems close to 50:50 on this.  That is, do you prefer `if (!ptr) {}` or `if (ptr == NULL)`?

**Comment by bel2125** [comment]
What do you mean by "mixed intent"?

If I understood it correctly, the 'AlwaysBreakAfterDefinitionReturnType' will always put the return type of a function in a single line. This is currently only the case if the line would be too long otherwise. The current style would allow 
```
int foo(int x, int y);
```
The new one one be
```
int
foo(int x, int y)
```
Or even
```
int
foo(
    int x,
    int y
    )
```
which seems unnecessary long. The current style will break only, if the line would be too long otherwise.

Regarding `if (!ptr)` vs. `if (ptr == NULL)`, I tend to prefer `if (ptr == NULL)`, but do not consider this as a strong requirement. I probably use both styles myself. Patches from other authors may also use one style or the other - I never rejected a patch because of code formatting reasons, and I never will (unless someone obviously intends to abuse this practice by sending [ioccc style code](http://www.ioccc.org/)).
Before I added the .clang-format file, some months ago, there were no source code formatting rules at all - so the style was even more mixed. Does clang-format support converting between `if (!ptr)` and `if (ptr == NULL)`?

**Comment by Parakleta** [comment]
Mixed **indent** means you currently have a mixture of tabs and spaces (tabs for indentation, spaces for alignment when wrapped) in the whitespace of the document.

The "break after return type" only applies to definitions and not declarations.  This is generally considered good practice because it makes it easy to search for the definition of a function by searching for the name at the start of the line.

This means that
```
int foo(int x, int y);
```
says the same, but
```
int foo(int x, int y)
{
    return x + y;
}
```
would become
```
int
foo(int x, int y)
{
    return x + y;
}
```
This then allows a search for the regular expression `^foo(` to find the definition of the `foo` function.  This style is currently mixed in the project, some functions use the former and some the latter (although my recollection is more of the latter).

clang-format cannot change syntax, only whitespace, so it cannot do anything about `!ptr` and `ptr == NULL`.  I used to prefer `ptr == NULL` because it was more "semantic", then I prefered `!ptr` because it was quicker and easier to type and read and was recognisable once you get used to it.  Then I had a problem with an embedded processor which had 96KiB of address space with only 16 bit pointers and 0 became a legal address (if it was extended with a high bit) and I had to define NULL as `((void *)-1)` (the address of the reset vector).  I think overall I still prefer `!ptr` because it's simpler and the '!' at the start immediately tells you it's the "bad" case.

I'm quite keen to refactor this code but refactoring tends to touch the code in a lot of places and if the style isn't consistent when you start it can make the patches that little bit messier and it can get a bit hard to follow what's going on.  This is especially the case when trying to factor out some code which is replicated but with different styles.

All of the other changes I'm happy to do as pull requests but the whitespace change in `.clang_format` will generate a diff which touches basically every line so I can't really submit that as a pull request.  The change to the "break after return type" could potentially be rolled into the same diff or I can submit that as a pull request.  The `!ptr` vs `ptr == NULL` I'll do as a pull request but I wanted to know if you had a preference before I did the work.  If you don't mind I think I'd rather do `!ptr`.

**Comment by bel2125** [comment]
Many years ago, long before Clang has been developed, I worked as a trainer for students who had to learn C - some of them, electrical engineers, who never did any programming before, and obviously just wanted to somehow pass their mandatory C course and never do any programming again. The only enforceable requirement was that their example programs somehow do approximately at least a part of the example specification.
Regarding "universally considered bad coding style": I could give A LOT of examples from that time - and once you'd seen them, an arbitrary mixture of spaces and tabs will not let you bat an eyelid.

I'm rather immune to bad code formatting.
Still, since now there are tools like .clang-format, of course one should use it.

Anyway let's have a look at the specific format options.

**Comment by bel2125** [comment]
"break after return type" seems reasonable, if it only affects the definitions.

I think I should do the .clang-format change before you factor out some `ssl.inl` (or whatever) file.

**Comment by bel2125** [comment]
I still do not really understand the problem with mixed intent:

A leading TAB will do the main indent every line, depending on the number of `{`.
The SPACE will do the vertical alignment, e.g. if a line needs to start in the 27th column to fit the position of the line above.

This seems quite natural to me - Why do you think this could be a problem?

**Comment by bel2125** [comment]
Regarding `!ptr` vs. `ptr == NULL` - let's just continue the current practice: everyone should do as he/she considers the best.

**Comment by bel2125** [comment]
Regarding all other changes to .clang-format: 
Do you have a suggested .clang-format file?

**Comment by Parakleta** [comment]
I used to be head tutor for a course at university which included a component of "coding standards".  I had to develop the coding standards and I've had to mark hundreds of assignments with a portion of their mark based on their compliance to the standards.  The golden rule above all else was consistency.  I guess I have the opposite situation to you, instead of being innoculated against coding style I now have this reflex to reach for a red pen. :)

The mixed indent a neat solution from a technical point of view, people can view the indent level based on their own preferences by changing the tab-stop setting and the wrapped code still lines up.  The issue is that I'm not aware of any editors which automatically support this format (certainly vim doesn't) which means you either have to carefully manually enter the mixed indent or you always have to post-process everything with clang-format before checking in.

The advantage of the mixed indent (that the level can be changed) has the downside that it messes with the 80 column limit and so in viewing the right margin becomes ragged when changing the ident.  Additionally, people who specifically want to change the indent for viewing or printing could just ask clang-format to do it for them and this will re-wrap the right margin correctly.

The advantage of not having mixed indent is that you can trivially add editor mode-lines so that vim and emacs (and maybe others) automatically generate the correct indent, simplifying the compliance with the required format.

I guess I see the difference being that without mixed indent people can be expected to generate the correct formatting and `clang-format` can be used as a verification step only (i.e. in the git pre-commit hooks), whereas with the mixed ident `clang-format` is basically required before commit to update the indentation to the complying style.

I'll propose a `.clang-format` file later today.

**Comment by bel2125** [comment]
I just tried to change the tab size with Visual Studio and with Notepad++ (both on Windows). They both use a default tab size of 4. If I change it to 2, 8, 7, .. it looks good - neither the `{` indent nor the vertical alignment are broken, neither with Notepad++ nor with Visual Studio.
I will test gedit on Linux and QtDesigner tomorrow, maybe also Eclipse/cdt. But I think all current IDEs do have some ways to auto-format the code this way - at least by customizing the menu to add an entry to call a format script.
Probably there is a way to let git handle this code formatting automatically.

**Comment by Parakleta** [comment]
On further research maybe my views on mixed indent are influenced by older unix conventions.  I am used to working with older automation tools which just didn't support mixed indent.  It seems on windows where the automation tools didn't exist there has been more acceptance of the mixed indent formatting, and now in the last 3-4 years across other domains as the automation tools have become more powerful the mixed indent is becoming more popular.

Maybe just chalk this up to me being old fashioned unix nerd and I'll update/customise my tools to suit the new style.

You can close this out if you like.  I'll submit the "break after return type" as a pull request later with some other minor fixes.

**Comment by bel2125** [comment]
OK. So let's stay with the mixed indent.

I would prefer if you send me a proposal of a .clang_format file (either just post it here, or as a pull request for this file only), as you said [above](https://github.com/civetweb/civetweb/issues/207#issuecomment-145293645). You can change the "break after return type" as we discussed - do you have any other suggestions? I would make the changes within a day or two. It's maybe easier to do for me, and to coordinate it with some commits (parts of a larger work) that are not pushed yet.

**Comment by Parakleta** [comment]
I've made a [gist](https://gist.github.com/Parakleta/cf33dfbf5b605a7cea0d) with a full `.clang_format` file as well as just an updated version of the existing one.

The lines that are commented out are those that just express the default settings (so the uncommented lines show the differences from the LLVM format).

The specific changes from the existing `.clang_format` are as follows:

* `AccessModifierOffset` - Removed since it's not applicable to C code.  For C++ code the default value de-indents the modifiers which I prefer.  This change is unimportant.
* `AlwaysBreakAfterDefinitionReturnType: true` - As Discussed.
* `BreakBeforeBinaryOperators: NonAssignment` - I find this style much easier to read because the operator is attached to the added information.  Consider reading "a, or b" versus "a or, b".  It also makes better diffs when adding or removing conditions in most cases because the operator and condition are added/removed as a single line.  This does make the case where a long function call is post-tested against a condition (i.e. `if (func(many_arguments) == result)` look a little ugly but this makes the test really stand out which I think is much safer.  Switching to pre-testing (i.e. `if (result == func(many_arguments))` can resolve the formatting if it's considered too ugly.
* `PenaltyBreakBeforeFirstCallParameter: 100` - This is effectively the argument equivalent of `AllowAllParametersOfDeclarationOnNextLine: false`.  It works together with `BinPackArguments: false` to generate equivalent formatting.  I'm not sure how the penalty system works but this value must be in the range [81..130] to get the correct behaviour.  `<=80` and nothing happens, `>130` and it messes up the formatting of struct initialisers.

**Comment by bel2125** [comment]
The option from the [gist](https://gist.github.com/Parakleta/cf33dfbf5b605a7cea0d) above does not seem to work:

```
D:\civetweb\civetweb>clang-format -i src/civetweb.c
YAML:19:39: error: invalid boolean
AlwaysBreakAfterDefinitionReturnType: TopLevel
                                      ^~~~~~~~
Error reading D:\civetweb\civetweb\.clang-format: invalid argument
Can't find usable .clang-format, using LLVM style

```

**Comment by bel2125** [comment]
The new format has been applied.

**Comment by Parakleta** [comment]
Looks good.  I missed the change in the 'break after definition' parameters between 3.6 and 3.7.  You're obviously using 3.6 and I've got 3.7 on my system.  `true` is equivalent to the new `All` definition which is less good for local function definitions or class methods, but I can't see that being an issue in pure C code.

**Comment by bel2125** [comment]
I usually do not upgrade my build systems for a couple of years.
A current minimum requirement is Visual Studio 2010 for Windows, and other compilers of ~5 years should still work. This makes it easier to support embedded systems, since some of them use also support only somewhat old compilers. It think I will stay with clang 3.6 also for a couple of years now.

**Comment by Parakleta** [comment]
LLVM v3.4 is the default system compiler on FreeBSD but it wasn't compatible with the clang-format file and since I didn't know which version you were using v3.7 was just the 'next cab off the rank'.  I've had to downgrade to v3.6 because even with an identical rule file v3.7 formats the file differently.  It may be worth adding a comment somewhere to say which version you're targeting.

**Comment by bel2125** [comment]
I did not know the clang-format version would be that important. I thought every new version could just add a few more options for fine-fine-tuning, with the default options producing the same output as the version before. This is obviously not true If identical rules for 3.6 and 3.7 do not produce the same output.
If I would have known, I would have spent some time to carefully select a version.
Some time ago I took the version that was available some for Windows, and this happened to be 3.6.0. Later I installed the same version on Linux as well.
But now this version is installed on all of my build systems.

Is there some way to specify a version in the .clang-format file - a style version, required minimum version, maximum version, required version .. something checked by clang-format, that will lead to a warning or error?

**Comment by Parakleta** [comment]
I haven't seen anything documented, but I can peek in the source and see if I can find anything.  Failing that using a build script that tests the version before reformatting would work (this is worth having anyway to keep track of which files should be managed).

**Comment by Parakleta** [comment]
I checked the source, there's no hidden options.  You'll have to manage this with the build system, but this is probably the best option anyway.  I can't help with that since I know nothing about CMake.

You may want to look at the scripts in [the clang repository](https://github.com/llvm-mirror/clang/tree/master/tools/clang-format) to see if any of those are helpful to integrate with the build system or git hooks.

**Comment by bel2125** [comment]
I think this issue can be closed now. 
If you disagree, feel free to open it again.
